src/storage/txn/commands/prewrite.rs

```diff
@@ -433,6 +433,29 @@ impl<K: PrewriteKind> Prewriter<K> {
         let mut final_min_commit_ts = TimeStamp::zero();
         let mut locks = Vec::new();
 
+        // Furthur check whether the prewrited transaction has been committed
+        // when encountering a WriteConflict or PessimisticLockNotFound error.
+        // This extra check manages to make prewrite idempotent after the transaction
+        // was committed.
+        // Note that this check cannot fully guarantee idempotence because an MVCC
+        // GC can remove the old committed records, then we cannot determine
+        // whether the transaction has been committed, so the error is still returned.
+        fn check_committed_record_on_err(
+            prewrite_result: MvccResult<(TimeStamp, OldValue)>,
+            reader: &mut SnapshotReader<impl Snapshot>,
+            key: &Key,
+        ) -> Result<(Vec<std::result::Result<(), StorageError>>, TimeStamp)> {
+            if let TxnCommitRecord::SingleRecord { commit_ts, .. } =
+                reader.get_txn_commit_record(key)?
```

Comment:
It might not matter much though, when the cursor is already pointing to some write record, it seems that calling `get_txn_commit_record` will seek again to the latest record, which I think is not necessary.

Reply:
I try the optimization like this: https://github.com/tikv/tikv/commit/ceab4385e400213f7e99893951b13836c816ee7c.

But finally I feel not so good about this optimization because it is based on a precondition: when the error occurs, the write cursor must not go beyond the commit record we want to find.

I'm afraid the precondition is broken some day in the future and we could not realize it. So I hesitate whether to add this optimizatin.

Reply:
I think it should call some `next`s not seek. You can confirm it from the metrics.

Reply:
Yes it calls some next but the first seek is actually a waste. But I want to keep the waste for safety and simplicity now.